src/H5Dmpio.c

```diff
@@ -1867,15 +1987,42 @@ H5D__link_chunk_filtered_collective_io(H5D_io_info_t *io_info, H5D_dset_io_info_
         H5MM_free(chunk_list.chunk_infos);
     } /* end if */
 
-    if (rank_chunks_assigned_map)
-        H5MM_free(rank_chunks_assigned_map);
+    /* Free resources used by cached dataset info */
+    if ((num_dset_infos == 1) && (chunk_list.dset_info.single_dset_info)) {
+        H5D_mpio_filtered_dset_info_t *curr_dset_info = chunk_list.dset_info.single_dset_info;
+
+        if (curr_dset_info->fb_info_init && H5D__fill_term(&curr_dset_info->fb_info) < 0)
+            HDONE_ERROR(H5E_DATASET, H5E_CANTFREE, FAIL, "can't release fill buffer info");
+        if (curr_dset_info->fill_space && H5S_close(curr_dset_info->fill_space) < 0)
+            HDONE_ERROR(H5E_DATASET, H5E_CLOSEERROR, FAIL, "can't close fill space");
+
+        H5MM_free(chunk_list.dset_info.single_dset_info);
+        chunk_list.dset_info.single_dset_info = NULL;
+    }
+    else if ((num_dset_infos > 1) && (chunk_list.dset_info.dset_info_hash_table)) {
+        H5D_mpio_filtered_dset_info_t *curr_dset_info;
+        H5D_mpio_filtered_dset_info_t *tmp;
+
+        HASH_ITER(hh, chunk_list.dset_info.dset_info_hash_table, curr_dset_info, tmp)
+        {
+            HASH_DELETE(hh, chunk_list.dset_info.dset_info_hash_table, curr_dset_info);
+
+            if (curr_dset_info->fb_info_init && H5D__fill_term(&curr_dset_info->fb_info) < 0)
+                HDONE_ERROR(H5E_DATASET, H5E_CANTFREE, FAIL, "can't release fill buffer info");
+            if (curr_dset_info->fill_space && H5S_close(curr_dset_info->fill_space) < 0)
+                HDONE_ERROR(H5E_DATASET, H5E_CLOSEERROR, FAIL, "can't close fill space");
+
+            H5MM_free(curr_dset_info);
+            curr_dset_info = NULL;
+        }
+    }
```

Comment:
Cleanup for the dataset info hash table occurs at the top-level in the I/O routines now rather than in each sub-function that was creating/freeing the hash table each time